### ISSUE TITLE
[Feature Request] Make tiling window resizing methodology more intuitive/"natural"

### ISSUE THREAD

**Issue by kmandarin-orange** [pop-os/cosmic-epoch#1166]
The tiling experience in Cosmic has been great so far, but I've noticed one confusing annoyance: it feels and unnatural to manage window resizing with a grow/shrink dichotomy when they're logically the exact same action: pushing a window's edge in one of two directions. I'll add some detail on problems with the current scheme, an alternative solution, and the pros/cons of said solution.

### Problems with Current Scheme

  - There are two logical "layers" to resizing: choosing to "grow" or "shrink" a window. This adds cognitive load to the concept of resizing by requiring users to make an extra choice about *how* to resize
  - All four edges of the window are modifiable at once, but rarely do users need to resize a window in all four directions
    - Additionally, it's rarely an option to edit all four edges of a window, since often the window needing resizing has one of its edges on the screen border
  - Splitting the problem into "growing" and "shrinking" adds a logical division that requires users to back-translate the UX flow to their desired action. Most people[^1] think about resizing windows by making two decisions: choose an edge to modify, and choose a direction to move it in. This logic is almost the reverse of the current implementation, increasing the cognitive effort to do the common task of window resizing

### Possible Solution

  1. Change to a directional scheme that only affects two window edges: a horizontal and vertical edge
  2. Default to selecting the bottom & right edges (unless one of those edges is against the screen edge/otherwise unable to resize, in which case the opposite edge is selected)
      1. Allows us reverse the default based on localization (e.g. bottom & left instead) for RTL languages/cultures that don't prefer LTR, but this would require some research to see if this actually makes sense to people
      1. Also allows using a single shortcut for resizing (or ideally, setting a modifier key to use with the `window focus` shortcuts)
          1. We could keep a second "invert" shortcut that would flip which edges are affected, but I'm not sure how useful that'd actually be to people, or if it'd add unnecessary complication to an intentionally simpler solution
  3. Use the up/down and left/right directional keys associated with focus direction to modify the bottom/top and right/left edges respectively. This simplifies the whole process down to moving the highlighted horizontal/vertical edges with their associated directional keys
  4. The help bubble that appears over the current window would show users something like:

```rust
"Resize {upper | lower} edge: {up/down keys} \n Resize {right | left} edge: {right/left keys}"
```

### Benefits 

  - *Much* more intuitive: there's a more obvious and direct connection between users' direction input and window movement, and there are simply less variables for them to worry about.
  - Allows a simpler keyboard shortcut trigger method: can be a single shortcut, or a chosen modifier key to use with the user's chosen `focus direction` shortcuts
  - More legible help bubble: since the text is no longer columned, the text has a much lower threshold for window width before it's broken into one-word lines or truncated
  - Shouldn't be any more complex in code than the current implementation, and could be more maintainable

### Drawbacks

  - Technically more limiting, though I'd argue the current design is limiting in terms of the occasional spatial/directional confusion it causes
  - Might require a couple more checks/logic for determining which edges to highlight
  - UI might need to be altered to reflect the new logic accurately
    - The current UI could be kept nearly identical, only needing to change the arrow icons on the 
    - Personally, I think it'd be ideal to have the focus outline *only* appear on the horizontal & vertical edge, but that'd require more design refactoring than is worth the minor benefit to keeping the appropriate edges emphasized

This is all a matter of preference of course, but I was wondering if others had felt a similar friction while using this feature. If so, exploring alternatives here would be interesting. If any of this was unclear, I can make some concept visuals to help.

[^1]: Source: "trust me bro" But really, the source is me and a few people I've talked to. I'm making some broad assumptions here, so hopefully this issue will surface other ways of thinking if my logic is off-base.

**Comment by nevdelap** [comment]
In my first few days of using cosmic, coming from i3, I started writing something like this but decided to leave it to think about more first. And every time I resize I feel that it is very unintuitive and feels very awkward. So I'm on the same page.

In i3, which I still use at work, and had used for many, many years at home as well, I have a binding for grow/shrink vertically using the up (grow) and down (shrink) keys and a binding for grow/shrink horizontally using the right (grow - because I use LTR languages) and left (shrink) keys. (if I were using an RTL language left/right would presumably make sense for grow/shrink horizontally) - acting on the currently active window. 

There is no overhead once you've gotten used to this because the behaviour is always the same. The window grows or shrinks in the right direction with the same keystroke. Which border moves just depends on which borders are against an edge of the screen, or is they aren't the window grows or shrinks proportionately in place, as the windows above and below or on either side change proportionately. Floating windows would work the same way, they just wouldn't affect other windows as they changed size.

UI wise there is no need for arrows to allow you what is happening because each arrow only does one thing always, and you don't have to think about it or and specifically don't need visual indications to help you think about what you are doing. 

So this probably counts as a suggestion for a second alternative, as well as a seconding. the motion there is something worth reconsidering in this specific area of the usability of cosmic.

**Comment by kmandarin-orange** [comment]
Fair point, it is all a matter of what people are used to. Glad to hear I'm not the only one noticing this!

I was afraid to pitch this as *yet another alternative option* for the devs to add to cosmic-settings though, since I assumed more people were used to the more common hyprland/i3/sway type of resizing, and didn't think the devs wanted to have another thing to maintain, hence the suggestion to replace the current method. 

If it turns out the split is more even than I expected, maybe it'd be worth adding an option to change the resize scheme for sure. Until then, it'll be interesting to see what the broader community thinks about how this should be improved.

**Comment by candsastle** [comment]
I definitely agree with changing things to be more intuitive and reduce the cognitive load when resizing windows.

What I would like to add is that I'm not sure how to use the resize shortcuts as they are currently. For example, if I bind "Super+H" to "Resize window inwards", and then if I press Super+H, it doesn't actually resize at all. It just goes into this resizing mode while I'm still holding down Super+H, where I then need to use the left arrow key to actually resize inwards, which I find not just confusing but also way too complicated for something like this. Am I just using this feature wrong or is that the actual intended workflow?

**Comment by kurisu-dotto-komu** [comment]
I would recommend just copying the previous Pop OS behavior and default keybindings?

This has the advantage of making things more intuitive for previous COSMIC users.

This was the only main stumbling block when migrating to the new COSMIC, as my muscle memory was used to Command + Enter to enter 'resize mode', and then just using arrow keys.

**Comment by victoreebel** [comment]
Yep, also feel this isn't hitting the spot intuition-wise. As you say there's just an additional layer of cognitive load vs. old Pop/tiling window manager behaviour. A secondary option would be welcomed, as well as bringing back the old resizing shortcuts.

**Comment by marcdimarco** [comment]
I use something like this in Sway and having similar functionality in Cosmic would be much more intuitive:

```#Resizing containers:
# Grow in direction (with fallback to shrink opposite side)
bindsym $mod+Ctrl+Shift+Right exec swaymsg resize grow right 20px || swaymsg resize shrink left 20px
bindsym $mod+Ctrl+Shift+Up exec swaymsg resize grow up 20px || swaymsg resize shrink down 20px
bindsym $mod+Ctrl+Shift+Down exec swaymsg resize grow down 20px || swaymsg resize shrink up 20px
bindsym $mod+Ctrl+Shift+Left exec swaymsg resize grow left 20px || swaymsg resize shrink right 20px
